### PR TITLE
Fix FollowingEngineTests#testOptimizeMultipleVersions

### LINKED ISSUE #72527: [CI] FollowingEngineTests testOptimizeMultipleVersions failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/ul4broxaituxw/tests/:x-pack:plugin:ccr:test/org.elasticsearch.xpack.ccr.index.engine.FollowingEngineTests/testOptimizeMultipleVersions

**Reproduction line:**
`./gradlew ':x-pack:plugin:ccr:test' --tests "org.elasticsearch.xpack.ccr.index.engine.FollowingEngineTests.testOptimizeMultipleVersions" -Dtests.seed=D5DEA1124A7EA74B -Dtests.locale=bg-BG -Dtests.timezone=Mexico/BajaNorte -Druntime.java=11 -Dtests.fips.enabled=true`

**Applicable branches:**
master, 7.x

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.ccr.index.engine.FollowingEngineTests&tests.test=testOptimizeMultipleVersions

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=153, name=Thread-115, state=RUNNABLE, group=TGRP-FollowingEngineTests]

  at __randomizedtesting.SeedInfo.seed([D5DEA1124A7EA74B:4EA15F6C2F3BA23]:0)

  Caused by: java.lang.AssertionError: maxSeqNoOfUpdates is not advanced local_checkpoint=80 msu=80 seq_no=81

    at __randomizedtesting.SeedInfo.seed([D5DEA1124A7EA74B]:0)
    at org.elasticsearch.xpack.ccr.index.engine.FollowingEngine.advanceMaxSeqNoOfUpdatesOrDeletesOnPrimary(FollowingEngine.java:125)
    at org.elasticsearch.index.engine.InternalEngine.index(InternalEngine.java:895)
    at org.elasticsearch.index.engine.EngineTestCase.applyOperation(EngineTestCase.java:1004)
    at org.elasticsearch.xpack.ccr.index.engine.FollowingEngineTests.fetchOperations(FollowingEngineTests.java:534)
    at org.elasticsearch.xpack.ccr.index.engine.FollowingEngineTests.lambda$runFollowTest$13(FollowingEngineTests.java:465)
    at java.lang.Thread.run(Thread.java:829)

```

### PR DESCRIPTION
**Pull request by fcofdez**
In certain concurrent indexing scenarios where there are deletes
executed and then a new indexing operation, the following engine
considers those as updates breaking one of the assumed invariants.

Closes #72527

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java [hunk 1]
```diff
@@ -119,13 +118,15 @@ protected long generateSeqNoForOperationOnPrimary(final Operation operation) {
 
     @Override
     protected void advanceMaxSeqNoOfUpdatesOrDeletesOnPrimary(long seqNo) {
-        if (Assertions.ENABLED) {
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java:122]: I wonder if we can keep the assertion for deletes and skip it for updates? You would need to split the method into two, but at least we then still have some of the assertion left.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineTests.java [hunk 2]
```diff
@@ -450,6 +461,63 @@ public void testOptimizeSingleDocConcurrently() throws Exception {
         });
     }
 
+    public void testConcurrentUpdateOperationsWithDeletesCanAdvanceMaxSeqNoOfUpdates() throws Exception {
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineTests.java:464]: The test here does not fail for me (without the `FollowingEngine` change), ran it 1000 times.

Can we maybe override the `advanceMaxSeqNoOfUpdatesOrDeletesOnPrimary` (and possibly other methods) in order to synchronize at the right time between the threads?

**Review by henningandersen — COMMENTED** [review]
I left a few comments, mainly on the test.

**Comment by fcofdez** [comment]
Sorry, I didn't have the chance to look into this until now. I've added a synchronization point between the test threads that should make the issue easier to reproduce, before it took > 1000 runs to reproduce.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java [hunk 3]
```diff
@@ -118,14 +118,27 @@ protected long generateSeqNoForOperationOnPrimary(final Operation operation) {
     }
 
     @Override
-    protected void advanceMaxSeqNoOfUpdatesOrDeletesOnPrimary(long seqNo) {
+    protected void advanceMaxSeqNoOfDeleteOnPrimary(long seqNo) {
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java:121]: I think we should keep this in plural, i.e., advanceMaxSeqNoOfDeletesOnPrimary

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java [hunk 4]
```diff
@@ -118,14 +118,27 @@ protected long generateSeqNoForOperationOnPrimary(final Operation operation) {
...
             assert localCheckpoint < maxSeqNoOfUpdates || maxSeqNoOfUpdates >= seqNo :
                 "maxSeqNoOfUpdates is not advanced local_checkpoint=" + localCheckpoint + " msu=" + maxSeqNoOfUpdates + " seq_no=" + seqNo;
         }
-        super.advanceMaxSeqNoOfUpdatesOrDeletesOnPrimary(seqNo); // extra safe in production code
+
+        super.advanceMaxSeqNoOfDeleteOnPrimary(seqNo);
+    }
+
+    @Override
+    protected void advanceMaxSeqNoOfUpdateOnPrimary(long seqNo) {
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java:133]: Also plural here for consistency.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java [hunk 5]
```diff
@@ -118,14 +118,27 @@ protected long generateSeqNoForOperationOnPrimary(final Operation operation) {
...
+        // In some scenarios it is possible to advance maxSeqNoOfUpdatesOrDeletes over the leader
+        // maxSeqNoOfUpdatesOrDeletes, since in this engine (effectively it is a replica) we don't check if the previous version
+        // was a delete and it's possible to consider it as an update, advancing the max sequence number over the leader
+        // maxSeqNoOfUpdatesOrDeletes.
+        // The goal of this marker it's just an optimization and it won't affect the correctness or durability of the indexed documents.
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngine.java:138]: A wrong number could lead to incorrect behavior so I suggest to rephrase this to something like:
// We conservatively advance the seqno in this case, accepting a minor performance hit in this edge case.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by fcofdez** [comment]
@elasticmachine update branch
